### ISSUE TITLE
Clarify Result<Stream> return type option for Subscription methods

### ISSUE THREAD

**Issue by komputerwiz** [async-graphql/async-graphql#1380]
The [documentation for Subscriptions](https://async-graphql.github.io/async-graphql/en/subscription.html) states: the "resolver function always returns a `Stream` or `Result<Stream>`." There is an example on the page that demonstrates the former, but I cannot find any examples demonstrating the latter, and I am having a lot of trouble figuring out how to do it.

I will admit that I am still relatively new to Rust and my knowledge in the use of traits is a bit sparse, but here is my specific issue...

I have the following subscription that currently operates as expected:

```rust
//! file: src/graphql/subscription.rs
use async_grpahql::{Context, Subscription};
use futures_core::Stream;
use tokio::sync::broadcast::Sender;

use crate::event::AppEvent;
use crate::graphql::model::Document;

pub struct Subscription;

#[Subscription]
impl Subscription {
    async fn document_create<'ctx>(&self, ctx: &Context<'ctx>) -> impl Stream<Item = Document> {
        // I would like for this to not panic if it fails
        let event_bus = ctx.data::<Sender<AppEvent>>().unwrap();

        BroadcastStream::new(event_bus.subscribe())
            .filter_map(|event| match event {
                Ok(AppEvent::DocumentCreated(doc)) => doc.try_into().ok(),
                _ => None,
            })
    }
}
```

However, the following does not compile:

```rust
use async_graphql::Result as GqlResult;

async fn document_create<'ctx>(&self, ctx: &Context<'ctx>) -> GqlResult<impl Stream<Item = Document>> {
    let event_bus = ctx.data::<Sender<AppEvent>>()?;

    let stream = BroadcastStream::new(event_bus.subscribe())
        .filter_map(move |event| match event {
            Ok(AppEvent::DocumentCreated(doc)) => doc.try_into().ok(),
            _ => None,
        });

    Ok(stream)
}
```

- ``error[E0667]: `impl Trait` is not allowed in path parameters``
- ``error[E0562]: `impl Trait` only allowed in function and inherent method return types, not in paths``

Nor does this:

```rust
async fn document_create<'ctx>(&self, ctx: &Context<'ctx>) -> GqlResult<dyn Stream<Item = Document>> {
    // ditto from above...
}
```

- ``error[E0308]: mismatched types ... expected `Result<dyn Stream<Item = Document>, Error>`, found `Result<FilterMap<..., ...>, ...>``
- ``error[E0277]: the size for values of type `dyn futures_core::Stream<Item = Document>` cannot be known at compilation time``

Nor this:

```rust
async fn document_create<'ctx>(&self, ctx: &Context<'ctx>) -> GqlResult<Box<dyn Stream<Item = Document>>> {
    // ...
    let stream: Box<dyn Stream<Item = Document>> = Box::new(/* ... */);

    Ok(stream)
}
```

- ``error[E0277]: `dyn futures_core::Stream<Item = GqlTag>` cannot be sent between threads safely``

I have tried several other variants have not been able to find a version that works. Any help would be greatly appreciated!

**Comment by lasantosr** [comment]
Same thing just happened to me.

It seems that the result must be named `Result` as it might be hardcoded somewhere, so `GqlResult` is not valid.
When using the right one the `impl` token from the return type is properly converted to `dyn`, but it doesn't perform that conversion if using another result name, which results in the error you see.

You have to remove the alias from the example, or if you're using other type just give it an alias to `Result`.

**Comment by sunli829** [comment]
Yes, there is an ugly hard coding here, the name must be `Result` or `FieldResult`.
